TheAlacrityCo/farm_link
# Jw da mh docker setup

# Description :: User Story

**Docker + CircleCI setup**
  * Add files: `Dockerfile`, `docker-compose.yml`, `entrypoint.sh`, `staging.rb`, `config.yml`, `staging.yml.enc`
  * Update files: `database.yml`, `webpacker.yml`, `Gemfile`, `development.rb`

This PR enables the Farm Link application to be run in `Docker` containers for test, development, and production. I took an SOA approach and used separate containers for the `Rails` app and `PostgreSQL` database, and they are linked through `Docker`. Compatibility with `CircleCI` is also included via the new `staging` environment, which is an environment specifically for CI/CD with `CircleCI`. The config for `CircleCI` is located in the root of the project `.circleci/config.yml`.

To build the `Docker` container and run it locally after merging into `master` do the following:
* Ensure your current terminal location is the `farm_link` project root directory
* `$ docker-compose build`
* `$ docker-compose up`
* Open a new terminal
  * `$ docker-compose run web rake db:create db:migrate`

To stop the application from running in `Docker`:
* `$ docker-compose down`

To run `RSpec` test suite:
* Ensure the application is running in `Docker`
  * To start the application use `$ docker-compose up`
* Open a new terminal
  * `$ docker-compose exec web rspec`
  * `RSpec` will run the test suite using the `Docker` containers

## Type of change


## Notes

Docker + Rails :: Setup
https://gist.github.com/iEv0lv3/17e33bc8d3ad4132d317ba5a1cf62d9b

CircleCI + Rails :: Setup
https://gist.github.com/iEv0lv3/0d3979312791f6a64440b3d320c0bea8

## RSpec results

```
............................

Finished in 1.05 seconds (files took 2.89 seconds to load)
28 examples, 0 failures

Coverage report generated for RSpec to /farm_link/coverage. 38 / 58 LOC (65.52%) covered.
```

## Rubocop results

```
Paste Rubocop results here
```
